### PR TITLE
COG pipeline

### PR DESCRIPTION
**Pull request by jterry64**
## Pull request checklist

Please check if your PR fulfills the following requirements:
- [X] Make sure you are requesting to pull a topic/feature/bugfix branch (right side). Don't request your master!
- [X] Make sure you are making a pull request against the develop branch (left side). Also you should start your branch off our develop.
- [X] Check the commit's or even all commits' message styles matches our requested structure.
- [X] Check your code additions will fail neither code linting checks nor unit test.

## Pull request type

 

 

Please check the type of change your PR introduces:
- [ ] Bugfix
- [X] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe): 

## What is the new behavior?
Create COG assets with specified projection and block size. This creates just one big COG file for all the times.

## Does this introduce a breaking change?

- [ ] Yes
- [X] No

### CODE REVIEW

File: app/models/pydantic/asset_metadata.py [hunk 1]
```diff
@@ -75,6 +75,12 @@ class RasterTileSetMetadata(AssetBase):
     resolution: Optional[int]
 
 
+class COGMetadata(AssetBase):
+    block_size: int
+    srid: int
+    resampling: int
```
Comment (dmannarino) [inline app/models/pydantic/asset_metadata.py:81]: Should these be 
block_size: TileBlockSize
srid: Srid
resampling: ResamplingMethod

File: batch/scripts/cogify.sh [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+ME=$(basename "$0")
+. get_arguments.sh "$@"
+
+set -x
+# download all GeoTiff files
+aws s3 cp --recursive --exclude "*" --include "*.tif" "${SRC}" .
+
+# combine to one big COG
+gdalwarp *.tif "${DATASET}_${VERSION}.tif" -r "${RESAMPLE}" -t_srs "${SRID}" -of COG -co BLOCKSIZE="${BLOCKSIZE}" -co NUM_THREADS=ALL_CPUS
```
[the remark this answers is not in the document]
Comment (jterry64) [reply batch/scripts/cogify.sh]: I like all these suggestions, will implement as I break up the gdalwarp into multiple shorter command

[the remark this answers is not in the document]
Comment (jterry64) [reply batch/scripts/cogify.sh]: To reproject a raster tile set to WGS84 without making a raster tile cache currently, would you just need to select a `zoom_{x}` grid?

File: app/models/pydantic/creation_options.py [hunk 3]
```diff
@@ -339,6 +340,26 @@ class RasterTileCacheCreationOptions(TileCacheBaseModel):
     )
 
 
+class COGCreationOptions(StrictBaseModel):
+    implementation: str = Field(
+        "default",
+        description="Name space to use for COG. "
+        "This will be part of the URI and will "
+        "allow to create multiple COGs per version,",
```
Comment (danscales) [inline app/models/pydantic/creation_options.py:348]: Optional: Change "... per version,"  to "... per version" (get rid of extra comma).  Also, maybe change:

"allow to create multiple COGs per version"

to

"allow creation of multiple COGs per version".

File: app/models/pydantic/jobs.py [hunk 4]
```diff
@@ -134,6 +137,19 @@ class PixETLJob(Job):
     attempt_duration_seconds = int(DEFAULT_JOB_DURATION * 1.5)
 
 
+class GDALCOGJob(Job):
+    """Use for raster transformations using GDAL Python docker in PixETL
```
Comment (danscales) [inline app/models/pydantic/jobs.py:141]: I think better described as:

"""Use for creating COG files using GDAL Python docker in PixETL queue"""

File: app/utils/path.py [hunk 5]
```diff
@@ -66,6 +66,8 @@ def get_asset_uri(
     uri_constructor: Dict[str, str] = {
         AssetType.dynamic_vector_tile_cache: f"{TILE_CACHE_URL}/{dataset}/{version}/dynamic/{{z}}/{{x}}/{{y}}.pbf",
         AssetType.static_vector_tile_cache: f"{TILE_CACHE_URL}/{dataset}/{version}/{implementation}/{{z}}/{{x}}/{{y}}.pbf",
+        # AssetType.cog: f"{TILE_CACHE_URL}/{dataset}/{version}/{implementation}/{{z}}/{{x}}/{{y}}.png",
```
Comment (danscales) [inline app/utils/path.py:69]: Remove this commented out path, unless you think it might be used later?

File: batch/scripts/get_arguments.sh [hunk 6]
```diff
@@ -172,6 +177,11 @@ do
       SKIP="TRUE"
       shift # past argument
       ;;
+      --srid)
+      SRID="$2"
+      shift # past argument
+      shift # past value
+      ;;
```
Comment (danscales) [inline batch/scripts/get_arguments.sh:184]: Was this new --srid ever used?  I don't see it used in cogify.sh

**Review by danscales — COMMENTED** [review]
Looks good to me!  Some minor suggestions.

File: batch/scripts/get_arguments.sh (hunk 6 above)
Comment (jterry64) [reply batch/scripts/get_arguments.sh:184]: We ended up removing it, good catch!

**Pull request merged**